aten/src/ATen/native/cuda/TensorTransformations.cu

```diff
@@ -0,0 +1,160 @@
+#include "ATen/NativeFunctions.h"
+#include "ATen/ATen.h"
+#include <algorithm>
+#include <sstream>
+
+#include "ATen/cuda/AccumulateType.cuh"
+#include "ATen/cuda/CUDATensorMethods.cuh"
+#include "ATen/cuda/CUDATypeConversion.cuh"
+
+
+namespace at {
+namespace native {
+
+// Map the index of an element in tensor from 1D to nD
+__device__ __forceinline__
+void oneD_to_nD(int64_t oneD_index, int64_t* shape_size, int64_t shape_len, int64_t* nD_index) {
+  int64_t res = oneD_index;
+  for (int i = 0; i < shape_len; i++) {
+    int64_t nD_i = oneD_index * shape_len + i;
+    nD_index[nD_i] = res / shape_size[i];
+    res = res % shape_size[i];
+  }
+}
+
+/*
+Map the index of an element in tensor from nD to 1D. A tensor is originally in nD shape,
+and 1D is the unfolded version of it (a vector).
+
+Example: given a 3D tensor
+[
+  [ [1,2], [3,4] ],
+  [ [5,6], [7,8] ]
+]
+
+Here element 3 has nD index = (0,1,0), and this corresponds to oneD index = 2
+*/
+__device__ __forceinline__
+int64_t nD_to_oneD(int64_t* nD_index, int64_t shape_len, int64_t* shape_size, int64_t src_oneD_index) {
+  int64_t dest_oneD_index = 0;
+  for (int i = 0; i < shape_len; i++) {
+    int64_t nD_i = src_oneD_index * shape_len + i;
+    dest_oneD_index += nD_index[nD_i] * shape_size[i];
+  }
+  return dest_oneD_index;
+}
+
+template <typename T>
+__global__
+void flip_cuda_kernel(T* in_t, T* out_t, int64_t N, int64_t* dims, int64_t* nD_index,
+  int64_t dims_len, int64_t* shape_size, int64_t* shape, int64_t shape_len) {
+
+  int64_t oneD_index = blockIdx.x * blockDim.x + threadIdx.x;
+  if (oneD_index >= N) {
+    return;
+  }
+
+  oneD_to_nD(oneD_index, shape_size, shape_len, nD_index);
+  // flip nD index along each desired dimension
+  for (int i = 0 ; i < dims_len; i++) {
+    int64_t d = dims[i];
+    int64_t nD_d = oneD_index * shape_len + d;
+    nD_index[nD_d] = shape[d]-1-nD_index[nD_d];
+  }
+  int64_t dest_oneD_index = nD_to_oneD(nD_index, shape_len, shape_size, oneD_index);
+  out_t[oneD_index] = in_t[dest_oneD_index];
+}
+
+Tensor flip_cuda(const Tensor& self, IntList dims) {
```

Comment:
I think we're making a hard assumption that `self` is contiguous. If this is the case, we should either be
1) operating on a contiguous tensor: use `self.contiguous()`
2) Making this algorithm work for non-contiguous tensors. I don't think this is easy, so we should probably not aim for this.